### PR TITLE
Reorganize base/string.jl, base/utf*, test/strings.jl, test/unicode.jl

### PR DESCRIPTION
**Pull request by ScottPJones**
The monolithic string.jl has been split up into several files,
and the test files in strings.jl and unicode.jl have been made
to correspond with the files of the same names in base.
This will prevent a lot of manual merging that was previously necessary.

This is a preface to trying to get the string coverage to 100%

The only changes are splitting the files up, placing them into folders, and adding a top file that
includes the files in the folder.

### CODE REVIEW

**Comment by tkelman** [comment]
The fact that things are in subfolders now is a little bit nicer than #11908, but I still don't think anyone is going to like the number of small files here. Base generally isn't organized as a single type per file. How about combining the various special string types together, into `SpecialStrings.jl` or something like that.

> This will prevent a lot of manual merging that was previously necessary.

Given how git works, I really don't see how this is the case. Yes you get silly little conflicts at the end of a file some times, those are the easy ones and really not an issue. Rebasing PR's uses up a bit of CI time, but it also tests things in a closer merge state than when they were originally opened. Adding more files seems to me that it would be more likely someone would modify the ends of multiple files, causing this to happen more often, not less.

**Comment by JeffBezanson** [comment]
This is better than the previous version.

Unrelated to this PR, but honestly I'd rather remove RepString, RevString, and RopeString.

I don't think we need utf/types.jl. Since we have utf16.jl and utf32.jl, we might as well put the corresponding type definitions in those files. However I wasn't paying attention when that file was introduced so I might be missing something.

Also looks like escape.jl and strio.jl could be combined.

**Comment by ScottPJones** [comment]
@tkelman About #11908, I was in the process of moving things to folders when it was closed on me (I'd already moved the test files, see last commit in #11908).  That hasn't been my experience with git.
I've been having to constantly rebase and do manual merging, on a daily basis, and not just for things at the end of a file.  Having more files that are logically separate means that it is more likely that two people working on strings will not conflict (one working on printing, the other on searching, for example).

@JeffBezanson "better than the previous version"  this is *exactly* what I was just about to push to the "previous version" when it got closed, and since I'm still fairly new with git and github, recreating it in a new branch ended up wasting a lot of my time and getting me frustrated.  Frustration has passed, just want to move on now.
`utf/types.jl` is because the type definitions are needed early on, before the stuff in `string` and `iobuffer`,
and they are needed before `utf/checkstring.jl`, which is needed before `utf/utf16.jl` and `utf/utf32.jl`
The utf changes was pure renaming in base (in test, it was splitting out so the test files match the source files).

You're correct about escape.jl and strio.jl.  I didn't see that at first because they were in different parts of the original string.jl.  I'm rerunning the unit tests with that change right now, will push shortly.

**Comment by ScottPJones** [comment]
@JeffBezanson OT about Rep/Rev/Rope, I'd already been bugging @tkelman about getting rid of Rope (it's not used anywhere, not in julia, not in any registered package, same thing is true of Rep).
He wanted me to move it into a package, but I haven't created any packages before, I can't create a package in JuliaLang, and I *don't* want to maintain code that I don't even think is needed.
`RevStrings` is used in a few places, once in `REPLCompletions`, once in `shell`, 5 times in `strings/search`. It isn't used in any registered packages either.
I'd be happy to simply delete Rep/Rope (there are no tests for them either, it would improve coverage stats also!).  What do you think?

**Comment by tkelman** [comment]
> That hasn't been my experience with git.
I've been having to constantly rebase and do manual merging, on a daily basis, and not just for things at the end of a file. Having more files that are logically separate means that it is more likely that two people working on strings will not conflict (one working on printing, the other on searching, for example).

This is because many of your proposed changes have been too big and contentious. The likelihood of needing to rebase to fix a conflict in an open PR is proportional to how many lines of code that PR touches, and has almost nothing to do with the number of files. This reorganization will not help reduce the number of conflicts in any way.

**Comment by ScottPJones** [comment]
@tkelman I've had that problem with rebasing even with small changes, it hasn't just been the children of #11004.  The problem usually has been just with the way `test/strings.jl` was just a dumping place for any string related tests.  This reorganization *will* help that quite a bit, if you look, the files in test *exactly* correspond to the files in base now, and in fact, the file `base/string.jl` is identical to `test/string.jl`.
(Previously splitting those up into even smaller logical units actually showed me that a number of things don't have *any* unit tests, something that has become hidden again by reducing the number of separate files in the strings folder!)

**Comment by tkelman** [comment]
So what's the general nature of the conflicts you've needed to resolve then?

I'm not opposed to reorganization, but we're going to need to find somewhere between master and a dozen small files. I know this sounds arbitrary, but can we find a way to do this while only adding around 5-6 new files?

**Comment by tkelman** [comment]
To clarify, by 5-6 new files I mean net additional files under base. Reorganizing the tests so the structure matches that in base is also good, I'm mostly focusing on converging to an agreeable structure of base here then the test changes can fall out of that.

**Comment by quinnj** [comment]
Scott, there are 55 other collaborators for JuliaLang that more or less have the ability to review and accept pull requests. When a PR is submitted that receives immediate feedback that it's not the right direction, and your response is a lengthy paragraph continuing to argue, it becomes a waste of everyone's time (particularly when dealing with more stylistic decisions where it's hard to be "objectively better"). You made no indication you were considering the suggested feedback, hence why the PR was closed. Note that you can continue to work on a branch, even if a PR is closed and it can be reopened later.

**Comment by ScottPJones** [comment]
Just stupid stuff, that perforce would have handled automatically, I'm not sure why git doesn't.
Sometimes it was because of mega changes outside of my control (Unions, Tuples, etc.),
other times it was just conflicts in test.
Since I do complete testing after any change (even one liners from rebasing), the turnaround is quite high for me.

**Comment by ScottPJones** [comment]
@quinnj One paragraph explaining why I'd structured things as I did, after only 4 comments from contributors, 1 that was just a comment about MATLAB, without any suggestions as what to do, and 1 was positive about the direction (just saying that the number was too much, which I was addressing when you closed it).  Giving somebody busy at least a day or two to respond would have been courteous.

**Comment by tkelman** [comment]
@ScottPJones much of the continuing friction is due to your responses to feedback. A simple "I have another change coming where I'm going to modify A,B,C, hold on" would have sufficed as opposed to https://github.com/JuliaLang/julia/pull/11908#issuecomment-116306354. Given the feedback you had received up to that point, I don't think you were going to convince anyone to like the level of granularity you had gone to.

**Comment by ScottPJones** [comment]
@tkelman I do think I've been *very* responsive to any constructive suggestions.  I push back to make sure that the suggestion really is OK, esp. if it goes directly against my own experience.

**Comment by ScottPJones** [comment]
Also, up til the point of my last commit (which, because of some screwup on my part due to my newness with git, and needing to rebase because of fixes in master that had prevented me from running my tests, had lost half of the change, where I moved the base files into the strings folder and merged a few, the PR was clearly marked as WIP, which means I'm still responding to the review suggestions)

**Comment by ScottPJones** [comment]
@tkelman In the comment you mentioned, I tried to explain my reasoning behind the way I was organizing things.  I even implied that I was going to change things further (but people had complained about doing more than one thing in a PR, so I said that it wasn't in that particular PR).
I think people need to be given at least time to respond to things from reviewers, and have the reviewers respond to that, before somebody pulls the plug on anything.  I've seen other complaints recently about things being merged or closed in less than a day, which is a big problem with a global community.

Is there anything else that needs to be done to this?  (@ninjin is also working on coverage for different parts of strings, so it would be good to get this in sooner rather than later, to avoid extra work from either of us)

**Comment by tkelman** [comment]
Yes, this needs to not add a dozen new small files to base. 5 or 6 would probably be okay.

**Comment by StefanKarpinski** [comment]
> Yes, this needs to not add a dozen new small files to base. 5 or 6 would probably be okay.

+1

**Comment by ScottPJones** [comment]
What change are you looking at?  This adds just 6 files in the folder strings, basic, search, parse, strutil, strio, and cstring.  Except for cstring, they are all fairly large, hundreds of lines.  cstring could probably go into basic, even though logically it didn't seem to belong.
There are completely separate types, SubStrings, RevStrings, RepStrings, and RopeStrings, which I've already commented on, two should simply be removed from base completely.  I was hoping that @JeffBezanson would respond about that.
Any other changes were just renaming files so that all of the utf* files are in the utf directory, so there's actually a net loss of files in Base, even with the files that I want to delete (RepStrings & RopeStrings)

**Comment by StefanKarpinski** [comment]
I think that if you combine the RopeString, RevString and RepString stuff into a single file named stringtypes.jl or something like that (on both the implementation and testing sides), then this would be ok. The logical separation is looking pretty reasonable at this point.

**Comment by ScottPJones** [comment]
What do you think about cstring (which is the only small one in the strings folder)?
On further thought, it really doesn't belong in the strings folder, maybe it should be added to either
`ascii.jl` or `utf/utf8.jl`, as it is a concrete type, and the rest of the strings folder is about `AbstractString`.
I'll merge the Rope/Rep/RevString (although I still want to hear from Jeff about killing Rope & Rep, if that could happen, it would just be SubString and RevString, which aren't small).

**Comment by ScottPJones** [comment]
Merging RopeString, RepString on the testing side is easy, there *is* no testing currently!
(another reason I wish them oblivion :grinning:)

**Comment by tkelman** [comment]
I'd also prefer `unicode` as a folder name and top-level base file, versus the context-free `utf`.

**Comment by StefanKarpinski** [comment]
If we get rid of one of those types, we'll probably get rid of all of them at the same time.

> I'd also prefer unicode as a folder name and top-level base file, versus the context-free utf.

+1

**Comment by ScottPJones** [comment]
OK, will rename from utf to unicode as well.

**Comment by ScottPJones** [comment]
Is there any way to inform git that I'm merging a file into another?  I think part of the difficulty of reviewing this refactoring is that it doesn't seem to handle merging or splitting nicely, but maybe I'm not using git correctly.  Thanks!

**Comment by StefanKarpinski** [comment]
No, the git model is that history is a series of tree snapshots – any notion of code motion is purely heuristic and only used for presentation or efficient storage.

**Comment by ScottPJones** [comment]
Ugh... I do miss perforce sometimes!

**Comment by JeffBezanson** [comment]
+1 for removing rope and rep

**Comment by ScottPJones** [comment]
Really, just kill them? :grinning:

**Comment by ScottPJones** [comment]
OK, now down to just 5 files under strings, all tests match the source files, and SubStrings.jl and RevStrings.jl only other new files in Base, net loss of a number of files directly in base (moved into unicode folder).

**Comment by ScottPJones** [comment]
The AppVeyor failure is because of some unrelated problem, likely due to JULIA_CPU_CORES == 1
(I've opened an Issue about those problems, which are preventing me from getting good coverage information currently)

Are there any other recommendations for this change?  Thanks!

**Comment by StefanKarpinski** [comment]
Those types will go in the long-awaited Stringnurok but not just yet. No point in breaking code in bits and pieces.

**Comment by ScottPJones** [comment]
What types are you talking about? 
This change is just making things easier to deal with, without changing any functionality.

**Comment by tkelman** [comment]
I still don't think RevStrings or SubStrings really deserve their own separate files.

**Comment by ScottPJones** [comment]
OK, I'll put them into a stringtypes.jl file.

**Comment by tkelman** [comment]
Deleting RopeStrings should not be done at this time, in this PR. We shouldn't just delete code until after we branch release-0.4.

**Comment by ScottPJones** [comment]
OK, I'll put it back, into strtypes.jl,  I did that because @JeffBezanson had said to remove it.

**Comment by ViralBShah** [comment]
Yes, we do need to remove it, but I have to think @tkelman is right not to do so this late in the 0.4 release cycle.

**Comment by ScottPJones** [comment]
OK, I have things all set to push with RopeString added back (now in stringtypes.jl (base & test)), but I'm waiting to see if the Appveyor and Travis-CI runs that are already in progress finish.
@ViralBShah I did check that absolutely no registered package uses RopeString, but I suppose somebody might have used it in private code.

**Comment by ScottPJones** [comment]
OK, all tests passed, I'm pushing the version with RopeString added back (which passed locally [there is only a single line of testing for RopeString]).   There are no substantive changes in this PR, it's just reorganization to make things more consistent and make life easier for those of us trying to increase testing coverage for strings.

**Comment by stevengj** [comment]
I agree that the `R*String` types should not be removed in this PR, just to keep it focused.  Deleting those should be a separate PR.  I wouldn't be opposed to merging that separate PR before 0.4, however, since those string types are used in literally *zero* registered packages at this point.

**Comment by ScottPJones** [comment]
It no longer removes anything, this is now strictly a cleaner restructuring of the files to help us split up the string coverage work.

**Comment by quinnj** [comment]
A few comments (this is shaping up nicely):
* let's move `strings/parse.jl` to just `base/parse.jl`; that functionality is more about parsing than strings
* rename `strio.jl` to just `io.jl`, `strutil.jl` to `util.jl`; they're already in the new `strings` directory
* it's probably ok where it is, but the `cmp` and `isequals` code feels like `basic.jl` rather than `util.jl` to me
* This may have been discussed elsewhere, but I would prefer `base/stringtypes.jl` to be `strings/types.jl`

**Comment by StefanKarpinski** [comment]
+1 to everything @quinnj said

**Comment by ViralBShah** [comment]
Also will need a squash.

**Comment by ScottPJones** [comment]
Into a single commit?  I'd been told to wait to do that until a PR was ready to be merged, if you think it's ready now, I'll squash however you want.

**Comment by ViralBShah** [comment]
It appears from the discussion that this is getting closer, so perhaps it would be a good time to do so.

**Comment by ScottPJones** [comment]
OK, will do.

**Comment by ScottPJones** [comment]
OK, all tests passed, and I think I've done everything people have asked for. Thanks for reviewing!

**Comment by stevengj** [comment]
LGTM.

**Comment by quinnj** [comment]
Looks like this needs a rebase, then LGTM

**Comment by ScottPJones** [comment]
All rebased, squashed to a single commit, passing all tests, :tada: Thanks again to all reviewers, it's greatly appreciated!

File: base/unicode/utf32.jl [hunk 1]
```diff
@@ -101,3 +101,37 @@ function map(f, s::UTF32String)
...
+end
+
+# pointer conversions of ASCII/UTF8/UTF16/UTF32 strings:
+pointer(x::Union{ByteString,UTF16String,UTF32String}) = pointer(x.data)
+pointer{T<:ByteString}(x::SubString{T}) = pointer(x.string.data) + x.offset
+pointer(x::ByteString, i::Integer) = pointer(x.data)+(i-1)
+pointer{T<:ByteString}(x::SubString{T}, i::Integer) = pointer(x.string.data) + x.offset + (i-1)
+pointer(x::Union{UTF16String,UTF32String}, i::Integer) = pointer(x)+(i-1)*sizeof(eltype(x.data))
+pointer{T<:Union{UTF16String,UTF32String}}(x::SubString{T}) = pointer(x.string.data) + x.offset*sizeof(eltype(x.data))
+pointer{T<:Union{UTF16String,UTF32String}}(x::SubString{T}, i::Integer) = pointer(x.string.data) + (x.offset + (i-1))*sizeof(eltype(x.data))
```
Comment (quinnj) [inline base/unicode/utf32.jl]: I think the `pointer` methods should be in `base/pointer.jl`. It seems a little obscure to put some of these methods in `utf32.jl`. Doesn't need to hold up the PR here, but a change we should probably make.

**Comment by tkelman** [comment]
The merge commit here froze during bootstrap https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.6734/job/41e1ujwp78e22yaw

the PR build passed, at https://github.com/JuliaLang/julia/commit/e3b1828 - not that many commits between the two

**Comment by kshyatt** [comment]
On my [latest Coveralls build](https://coveralls.io/builds/3047587) the string coverage is looking really great. Awesome work, @ScottPJones.

**Comment by malmaud** [comment]
This PR clobbered https://github.com/JuliaLang/julia/pull/11898, possibly other PRs that touched the string files.

**Comment by tkelman** [comment]
would've been ideal if there were a script or some other mapping to actually look at the diff piece by piece and really verify that this was strictly organizational at the time of merging

**Comment by yuyichao** [comment]
A Google search leads to [this](http://stackoverflow.com/questions/7759193/git-diff-renamed-file) and similar ones which suggests that `git diff -M/-C` should work.

**Pull request merged**